From b9e89a1437a02e76fc02426eb85a3f1950678b65 Mon Sep 17 00:00:00 2001
From: Christian Marangi <ansuelsmth@gmail.com>
Date: Sat, 15 Oct 2022 09:54:11 +0200
Subject: [PATCH] info: rework info_thrd to POSIX variant

sigtimedwait and sigwaitinfo is not POSIX and is not present on macos.

Rework the logic with gettimeofday to get the second interval and revert
to POSIX sigwait.

Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>
---
 squashfs-tools/info.c            | 27 ++++++++-------------------
 squashfs-tools/unsquashfs_info.c | 27 ++++++++-------------------
 2 files changed, 16 insertions(+), 38 deletions(-)

diff --git a/squashfs-tools/info.c b/squashfs-tools/info.c
index a34dcca..563bf63 100644
--- a/squashfs-tools/info.c
+++ b/squashfs-tools/info.c
@@ -144,7 +144,7 @@ static void dump_state()
 static void *info_thrd(void *arg)
 {
 	sigset_t sigmask;
-	struct timespec timespec = { .tv_sec = 1, .tv_nsec = 0 };
+	struct timeval old_time = {0}, cur_time = {0};
 	int sig, waiting = 0;
 
 	sigemptyset(&sigmask);
@@ -152,25 +152,12 @@ static void *info_thrd(void *arg)
 	sigaddset(&sigmask, SIGHUP);
 
 	while(1) {
-		if(waiting)
-			sig = sigtimedwait(&sigmask, NULL, &timespec);
-		else
-			sig = sigwaitinfo(&sigmask, NULL);
-
-		if(sig == -1) {
-			switch(errno) {
-			case EAGAIN:
-				/* interval timed out */
+		sigwait(&sigmask, &sig);
+
+		if(waiting) {
+			gettimeofday(&cur_time, NULL);
+			if (cur_time.tv_sec > old_time.tv_sec )
 				waiting = 0;
-				/* FALLTHROUGH */
-			case EINTR:
-				/* if waiting, the wait will be longer, but
-				   that's OK */
-				continue;
-			default:
-				BAD_ERROR("sigtimedwait/sigwaitinfo failed "
-					"because %s\n", strerror(errno));
-			}
 		}
 
 		if(sig == SIGQUIT && !waiting) {
@@ -181,6 +168,8 @@ static void *info_thrd(void *arg)
 			waiting = 1;
 		} else
 			dump_state();
+
+		gettimeofday(&old_time, NULL);
 	}
 }
 
diff --git a/squashfs-tools/unsquashfs_info.c b/squashfs-tools/unsquashfs_info.c
index e906eaf..6015634 100644
--- a/squashfs-tools/unsquashfs_info.c
+++ b/squashfs-tools/unsquashfs_info.c
@@ -96,7 +96,7 @@ void dump_state()
 void *info_thrd(void *arg)
 {
 	sigset_t sigmask;
-	struct timespec timespec = { .tv_sec = 1, .tv_nsec = 0 };
+	struct timeval old_time = {0}, cur_time = {0};
 	int sig, waiting = 0;
 
 	sigemptyset(&sigmask);
@@ -104,25 +104,12 @@ void *info_thrd(void *arg)
 	sigaddset(&sigmask, SIGHUP);
 
 	while(1) {
-		if(waiting)
-			sig = sigtimedwait(&sigmask, NULL, &timespec);
-		else
-			sig = sigwaitinfo(&sigmask, NULL);
-
-		if(sig == -1) {
-			switch(errno) {
-			case EAGAIN:
-				/* interval timed out */
+		sigwait(&sigmask, &sig);
+
+		if(waiting) {
+			gettimeofday(&cur_time, NULL);
+			if (cur_time.tv_sec > old_time.tv_sec )
 				waiting = 0;
-				/* FALLTHROUGH */
-			case EINTR:
-				/* if waiting, the wait will be longer, but
-				   that's OK */
-				continue;
-			default:
-				BAD_ERROR("sigtimedwait/sigwaitinfo failed "
-					"because %s\n", strerror(errno));
-			}
 		}
 
 		if(sig == SIGQUIT && !waiting) {
@@ -134,6 +121,8 @@ void *info_thrd(void *arg)
 			waiting = 1;
 		} else
 			dump_state();
+
+		gettimeofday(&old_time, NULL);
 	}
 }
 
-- 
2.37.2

